substrate/frame/revive/proc-macro/src/lib.rs

```diff
@@ -339,48 +343,61 @@ where
 	P: Iterator<Item = &'a std::boxed::Box<syn::Pat>> + Clone,
 	I: Iterator<Item = &'a std::boxed::Box<syn::Type>> + Clone,
 {
-	const ALLOWED_REGISTERS: u32 = 6;
-	let mut registers_used = 0;
-	let mut bindings = vec![];
-	for (idx, (name, ty)) in param_names.clone().zip(param_types.clone()).enumerate() {
+	const ALLOWED_REGISTERS: usize = 6;
+
+	// all of them take one register but we truncate them before passing into the function
+	// it is important to not allow any type which has illegal bit patterns like 'bool'
+	if !param_types.clone().all(|ty| {
 		let syn::Type::Path(path) = &**ty else {
 			panic!("Type needs to be path");
 		};
 		let Some(ident) = path.path.get_ident() else {
 			panic!("Type needs to be ident");
 		};
-		let size = if ident == "i8" ||
-			ident == "i16" ||
-			ident == "i32" ||
-			ident == "u8" ||
-			ident == "u16" ||
-			ident == "u32"
-		{
-			1
-		} else if ident == "i64" || ident == "u64" {
-			2
-		} else {
-			panic!("Pass by value only supports primitives");
-		};
-		registers_used += size;
-		if registers_used > ALLOWED_REGISTERS {
-			return quote! {
-				let (#( #param_names, )*): (#( #param_types, )*) = memory.read_as(__a0__)?;
-			}
-		}
-		let this_reg = quote::format_ident!("__a{}__", idx);
-		let next_reg = quote::format_ident!("__a{}__", idx + 1);
-		let binding = if size == 1 {
+		matches!(ident.to_string().as_ref(), "u8" | "u16" | "u32" | "u64")
+	}) {
+		panic!("Only primitive unsigned integers are allowed as arguments to syscalls");
+	}
+
+	// too many arguments: pass as pointer to a struct in memory
+	if param_names.clone().count() > ALLOWED_REGISTERS {
+		let fields = param_names.clone().zip(param_types.clone()).map(|(name, ty)| {
 			quote! {
-				let #name = #this_reg as #ty;
+				#name: #ty,
 			}
-		} else {
-			quote! {
-				let #name = (#this_reg as #ty) | ((#next_reg as #ty) << 32);
+		});
+		return quote! {
+			#[derive(Default)]
+			#[repr(C)]
+			struct Args {
+				#(#fields)*
 			}
-		};
-		bindings.push(binding);
+			let Args { #(#param_names,)* } = {
+				let len = ::core::mem::size_of::<Args>();
+				let mut args = Args::default();
+				let ptr = &mut args as *mut Args as *mut u8;
+				// Safety
+				// 1. The struct is initialized at all times.
+				// 2. We only allow primitive integers (no bools) as arguments so every bit pattern is safe.
+				// 3. The reference doesn't outlive the args field.
+				// 4. There is only the single reference to the args field.
+				// 5. The length of the generated slice is the same as the struct.
+				let reference = unsafe {
+					::core::slice::from_raw_parts_mut(ptr, len)
+				};
+				memory.read_into_buf(__a0__ as _, reference)?;
+				args
+			};
+		}
 	}
+
+	// otherwise: one argument per register
+	let bindings = param_names.zip(param_types).enumerate().map(|(idx, (name, ty))| {
+		let reg = quote::format_ident!("__a{}__", idx);
+		quote! {
+			let #name: #ty = #reg.try_into().map_err(|_| <Error<E::T>>::ArgumentTooLarge)?;
```

Comment:
Unfortunately we can't guard against this, it works against the [ABI spec](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/releases/download/v1.0/riscv-abi.pdf).

`integer scalars narrower than XLEN bits are widened according to the sign of their type up to 32 bits, then sign-extended to XLEN`

So looks like the only thing what we can do is to turn this into a documentation issue, i.e. stating that contract code is responsible to not supply pointer values above `u32::MAX` or this will silently overflow?

Reply:
Aye. Reverted the change.